tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -6017,6 +6017,180 @@ Status ConvertTopK(OpConverterParams* params) {
+  // Does the same conversion steps as ConvertDepthSpaceShuffle. The only
+  // modification is that we cannot compute the reshape dimensions from
+  // inputs.at(0).GetTrtDims() (as they have -1 values). Instead we use a shape
+  // layer and shape arithmetic to calculate the reshape dimensions.
+  const auto& inputs = params->inputs;
+  const auto& node_def = params->node_def;
+  if (params->use_implicit_batch) return errors::Internal("TRT6 required.");
+  TFAttrs attrs(node_def);
+  const int block_size = attrs.get<int64>("block_size");
+  if (block_size < 2) {
+    return errors::InvalidArgument("Block size must be 2 or greater, at ",
+                                   node_def.name());
+  }
+  const string data_format = attrs.get<string>("data_format");
+  if (data_format != "NCHW" && data_format != "NHWC") {
+    return errors::Unimplemented("Data format ", data_format,
+                                 " is not supported, at ", node_def.name());
+  }
+  nvinfer1::Dims dims = inputs.at(0).GetTrtDims();
+  if (dims.nbDims != 4) {
+    return errors::InvalidArgument("The input to ", node_def.op(),
+                                   " must be rank 4, at ", node_def.name());
+  }
+  const int channels_axis = data_format == "NCHW" ? 1 : 3;
+  const int h_axis = data_format == "NCHW" ? 2 : 1;
+  const int w_axis = data_format == "NCHW" ? 3 : 2;
+  if (node_def.op() == "DepthToSpace" && dims.d[channels_axis] != -1 &&
+      dims.d[channels_axis] % (block_size * block_size) != 0) {
+    return errors::InvalidArgument(
+        "Number of channels must be divisible by block_size*block_size, at ",
+        node_def.name());
+  }
+  if (node_def.op() == "SpaceToDepth" &&
+      ((dims.d[h_axis] != -1 && dims.d[h_axis] % block_size != 0) ||
+       (dims.d[w_axis] != -1 && dims.d[w_axis] % block_size != 0))) {
+    return errors::InvalidArgument(
+        "Width and height must be divisible by block_size, at ",
+        node_def.name());
+  }
+  if (params->validation_only) return Status::OK();
+
+  // Get shapes.
+  nvinfer1::ITensor* shape = params->converter->network()
+                                 ->addShape(*inputs.at(0).tensor())
+                                 ->getOutput(0);
+  nvinfer1::ITensor* batch_size =
+      params->converter->network()
+          ->addSlice(*shape, {1, {0}}, {1, {1}}, {1, {1}})
+          ->getOutput(0);
+  nvinfer1::ITensor* num_channels =
+      params->converter->network()
+          ->addSlice(*shape, {1, {channels_axis}}, {1, {1}}, {1, {1}})
+          ->getOutput(0);
+  nvinfer1::ITensor* h =
+      params->converter->network()
+          ->addSlice(*shape, {1, {h_axis}}, {1, {1}}, {1, {1}})
+          ->getOutput(0);
+  nvinfer1::ITensor* w =
+      params->converter->network()
+          ->addSlice(*shape, {1, {w_axis}}, {1, {1}}, {1, {1}})
+          ->getOutput(0);
+  nvinfer1::ITensor* r;
+  TF_RETURN_IF_ERROR(CreateScalarConstant(params, block_size, &r));
```

Comment:
I wonder if we can avoid having ConvertDepthSpaceShuffleDynamic, which duplicates a bunch of code from ConvertDepthSpaceShuffle. Instead, extend ConvertDepthSpaceShuffle to call a routine to "find/generate" batch_size/num_channels/h/w etc in the presence of dynamic shape for input(0).

Reply:
I removed the duplication, kept only the shape tensor arithmetic, and use it as a helper in the original converter.